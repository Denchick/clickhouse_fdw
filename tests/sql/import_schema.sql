CREATE EXTENSION clickhouse_fdw;
SET datestyle = 'ISO';
CREATE SERVER loopback FOREIGN DATA WRAPPER clickhouse_fdw
    OPTIONS(dbname 'regression', driver 'http');
CREATE SERVER loopback_bin FOREIGN DATA WRAPPER clickhouse_fdw
    OPTIONS(dbname 'regression', driver 'binary');
CREATE SCHEMA clickhouse;
CREATE SCHEMA clickhouse_bin;
CREATE SCHEMA clickhouse_limit;
CREATE SCHEMA clickhouse_except;
CREATE USER MAPPING FOR CURRENT_USER SERVER loopback;
CREATE USER MAPPING FOR CURRENT_USER SERVER loopback_bin;

SELECT clickhousedb_raw_query('DROP DATABASE IF EXISTS regression');
SELECT clickhousedb_raw_query('CREATE DATABASE regression');

-- integer types
SELECT clickhousedb_raw_query('CREATE TABLE regression.ints (
    c1 Int8, c2 Int16, c3 Int32, c4 Int64,
    c5 UInt8, c6 UInt16, c7 UInt32, c8 UInt64,
    c9 Float32, c10 Nullable(Float64)
) ENGINE = MergeTree PARTITION BY c1 ORDER BY (c1);
');
SELECT clickhousedb_raw_query('INSERT INTO regression.ints SELECT
    number, number + 1, number + 2, number + 3, number + 4, number + 5,
    number + 6, number + 7, number + 8.1, number + 9.2 FROM numbers(10);');
SELECT clickhousedb_raw_query('INSERT INTO regression.ints SELECT
    number, number + 1, number + 2, number + 3, number + 4, number + 5,
    number + 6, number + 7, number + 8.1, NULL FROM numbers(10, 2);');

SELECT clickhousedb_raw_query('CREATE TABLE regression.types (
    c1 Date, c2 DateTime, c3 String, c4 FixedString(5), c5 UUID,
    c6 Enum8(''one'' = 1, ''two'' = 2),
    c7 Enum16(''one'' = 1, ''two'' = 2, ''three'' = 3),
    c9 Nullable(FixedString(50)),
    c8 LowCardinality(String)
) ENGINE = MergeTree PARTITION BY c1 ORDER BY (c1);
');
SELECT clickhousedb_raw_query('INSERT INTO regression.types SELECT
    addDays(toDate(''1990-01-01''), number),
    addMinutes(addSeconds(addDays(toDateTime(''1990-01-01 10:00:00''), number), number), number),
    format(''number {0}'', toString(number)),
    format(''num {0}'', toString(number)),
    format(''f4bf890f-f9dc-4332-ad5c-0c18e73f28e{0}'', toString(number)),
    ''two'',
    ''three'',
    toString(number),
    format(''cardinal {0}'', toString(number))
    FROM numbers(10);');

-- array types
SELECT clickhousedb_raw_query('CREATE TABLE regression.arrays (
    c1 Array(Int), c2 Array(String)
) ENGINE = MergeTree PARTITION BY c1 ORDER BY (c1);
');
SELECT clickhousedb_raw_query('INSERT INTO regression.arrays SELECT
    [number, number + 1],
    [format(''num{0}'', toString(number)), format(''num{0}'', toString(number + 1))]
    FROM numbers(10);');

-- tuple
SELECT clickhousedb_raw_query('CREATE TABLE regression.tuples (
    c1 Int8,
    c2 Tuple(Int, String, Float32)
) ENGINE = MergeTree PARTITION BY c1 ORDER BY (c1);
');
SELECT clickhousedb_raw_query('INSERT INTO regression.tuples SELECT
    number,
    (number, toString(number), number + 1.0)
    FROM numbers(10);');

IMPORT FOREIGN SCHEMA "<does not matter>" FROM SERVER loopback INTO clickhouse;

\d+ clickhouse.ints;
\d+ clickhouse.types;
\d+ clickhouse.arrays;
\d+ clickhouse.tuples;

SELECT * FROM clickhouse.ints ORDER BY c1 DESC LIMIT 4;
SELECT * FROM clickhouse.types ORDER BY c1 LIMIT 2;
SELECT * FROM clickhouse.arrays ORDER BY c1 LIMIT 2;
SELECT * FROM clickhouse.tuples ORDER BY c1 LIMIT 2;

IMPORT FOREIGN SCHEMA "<does not matter>" FROM SERVER loopback_bin INTO clickhouse_bin;

\d+ clickhouse_bin.ints;
\d+ clickhouse_bin.types;
\d+ clickhouse_bin.arrays;
\d+ clickhouse_bin.tuples;

SELECT * FROM clickhouse_bin.ints ORDER BY c1 DESC LIMIT 4;
SELECT * FROM clickhouse_bin.types ORDER BY c1 LIMIT 2;
SELECT * FROM clickhouse_bin.arrays ORDER BY c1 LIMIT 2;
SELECT * FROM clickhouse_bin.tuples ORDER BY c1 LIMIT 2;

IMPORT FOREIGN SCHEMA "<does not matter>" LIMIT TO (ints, types) FROM SERVER loopback INTO clickhouse_limit;

\d+ clickhouse_limit.ints;
\d+ clickhouse_limit.types;
\d+ clickhouse_limit.arrays;
\d+ clickhouse_limit.tuples;

IMPORT FOREIGN SCHEMA "<does not matter>" EXCEPT (ints, types) FROM SERVER loopback INTO clickhouse_except;

\d+ clickhouse_except.ints;
\d+ clickhouse_except.types;
\d+ clickhouse_except.arrays;
\d+ clickhouse_except.tuples;

DROP USER MAPPING FOR CURRENT_USER SERVER loopback;
DROP USER MAPPING FOR CURRENT_USER SERVER loopback_bin;

SELECT clickhousedb_raw_query('DROP DATABASE regression');
DROP EXTENSION clickhouse_fdw CASCADE;
