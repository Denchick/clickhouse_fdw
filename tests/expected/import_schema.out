CREATE EXTENSION clickhouse_fdw;
SET datestyle = 'ISO';
CREATE SERVER loopback FOREIGN DATA WRAPPER clickhouse_fdw
    OPTIONS(dbname 'regression', driver 'http');
CREATE SERVER loopback_bin FOREIGN DATA WRAPPER clickhouse_fdw
    OPTIONS(dbname 'regression', driver 'binary');
CREATE SCHEMA clickhouse;
CREATE SCHEMA clickhouse_bin;
CREATE SCHEMA clickhouse_limit;
CREATE SCHEMA clickhouse_except;
CREATE USER MAPPING FOR CURRENT_USER SERVER loopback;
CREATE USER MAPPING FOR CURRENT_USER SERVER loopback_bin;
SELECT clickhousedb_raw_query('DROP DATABASE IF EXISTS regression');
 clickhousedb_raw_query 
------------------------
 
(1 row)

SELECT clickhousedb_raw_query('CREATE DATABASE regression');
 clickhousedb_raw_query 
------------------------
 
(1 row)

-- integer types
SELECT clickhousedb_raw_query('CREATE TABLE regression.ints (
    c1 Int8, c2 Int16, c3 Int32, c4 Int64,
    c5 UInt8, c6 UInt16, c7 UInt32, c8 UInt64,
    c9 Float32, c10 Nullable(Float64)
) ENGINE = MergeTree PARTITION BY c1 ORDER BY (c1);
');
 clickhousedb_raw_query 
------------------------
 
(1 row)

SELECT clickhousedb_raw_query('INSERT INTO regression.ints SELECT
    number, number + 1, number + 2, number + 3, number + 4, number + 5,
    number + 6, number + 7, number + 8.1, number + 9.2 FROM numbers(10);');
 clickhousedb_raw_query 
------------------------
 
(1 row)

SELECT clickhousedb_raw_query('INSERT INTO regression.ints SELECT
    number, number + 1, number + 2, number + 3, number + 4, number + 5,
    number + 6, number + 7, number + 8.1, NULL FROM numbers(10, 2);');
 clickhousedb_raw_query 
------------------------
 
(1 row)

SELECT clickhousedb_raw_query('CREATE TABLE regression.types (
    c1 Date, c2 DateTime, c3 String, c4 FixedString(5), c5 UUID,
    c6 Enum8(''one'' = 1, ''two'' = 2),
    c7 Enum16(''one'' = 1, ''two'' = 2, ''three'' = 3),
    c9 Nullable(FixedString(50)),
    c8 LowCardinality(String)
) ENGINE = MergeTree PARTITION BY c1 ORDER BY (c1);
');
 clickhousedb_raw_query 
------------------------
 
(1 row)

SELECT clickhousedb_raw_query('INSERT INTO regression.types SELECT
    addDays(toDate(''1990-01-01''), number),
    addMinutes(addSeconds(addDays(toDateTime(''1990-01-01 10:00:00''), number), number), number),
    format(''number {0}'', toString(number)),
    format(''num {0}'', toString(number)),
    format(''f4bf890f-f9dc-4332-ad5c-0c18e73f28e{0}'', toString(number)),
    ''two'',
    ''three'',
    toString(number),
    format(''cardinal {0}'', toString(number))
    FROM numbers(10);');
 clickhousedb_raw_query 
------------------------
 
(1 row)

-- array types
SELECT clickhousedb_raw_query('CREATE TABLE regression.arrays (
    c1 Array(Int), c2 Array(String)
) ENGINE = MergeTree PARTITION BY c1 ORDER BY (c1);
');
 clickhousedb_raw_query 
------------------------
 
(1 row)

SELECT clickhousedb_raw_query('INSERT INTO regression.arrays SELECT
    [number, number + 1],
    [format(''num{0}'', toString(number)), format(''num{0}'', toString(number + 1))]
    FROM numbers(10);');
 clickhousedb_raw_query 
------------------------
 
(1 row)

-- tuple
SELECT clickhousedb_raw_query('CREATE TABLE regression.tuples (
    c1 Int8,
    c2 Tuple(Int, String, Float32),
	c3 Nested(a Int, b Int),
	c4 Int16
) ENGINE = MergeTree PARTITION BY c1 ORDER BY (c1);
');
 clickhousedb_raw_query 
------------------------
 
(1 row)

SELECT clickhousedb_raw_query('INSERT INTO regression.tuples SELECT
    number,
    (number, toString(number), number + 1.0),
	[toInt32(number),1,1],
	[toInt32(number),2,2],
	toInt16(number)
    FROM numbers(10);');
 clickhousedb_raw_query 
------------------------
 
(1 row)

-- datetime64
SELECT clickhousedb_raw_query('CREATE TABLE regression.dt64 (
    c1 Int32,
    c2 DateTime64(6, ''Europe/Moscow''),
    c3 DateTime64(10, ''Europe/Berlin''),
	c4 DateTime64(6)
) ENGINE = MergeTree PARTITION BY c1 ORDER BY (c1);
');
 clickhousedb_raw_query 
------------------------
 
(1 row)

IMPORT FOREIGN SCHEMA "<does not matter>" FROM SERVER loopback INTO clickhouse;
NOTICE:  clickhouse_fdw: ClickHouse <Tuple> type was translated to <TEXT> type, please create composite type and alter the column if needed
\d+ clickhouse.ints;
                                         Foreign table "clickhouse.ints"
 Column |       Type       | Collation | Nullable | Default | FDW options | Storage | Stats target | Description 
--------+------------------+-----------+----------+---------+-------------+---------+--------------+-------------
 c1     | smallint         |           | not null |         |             | plain   |              | 
 c2     | smallint         |           | not null |         |             | plain   |              | 
 c3     | integer          |           | not null |         |             | plain   |              | 
 c4     | bigint           |           | not null |         |             | plain   |              | 
 c5     | smallint         |           | not null |         |             | plain   |              | 
 c6     | integer          |           | not null |         |             | plain   |              | 
 c7     | bigint           |           | not null |         |             | plain   |              | 
 c8     | bigint           |           | not null |         |             | plain   |              | 
 c9     | real             |           | not null |         |             | plain   |              | 
 c10    | double precision |           |          |         |             | plain   |              | 
Server: loopback
FDW options: (table_name 'ints', engine 'MergeTree')

\d+ clickhouse.types;
                                              Foreign table "clickhouse.types"
 Column |            Type             | Collation | Nullable | Default | FDW options | Storage  | Stats target | Description 
--------+-----------------------------+-----------+----------+---------+-------------+----------+--------------+-------------
 c1     | date                        |           | not null |         |             | plain    |              | 
 c2     | timestamp without time zone |           | not null |         |             | plain    |              | 
 c3     | text                        |           | not null |         |             | extended |              | 
 c4     | character varying(5)        |           | not null |         |             | extended |              | 
 c5     | uuid                        |           | not null |         |             | plain    |              | 
 c6     | text                        |           | not null |         |             | extended |              | 
 c7     | text                        |           | not null |         |             | extended |              | 
 c9     | character varying(50)       |           |          |         |             | extended |              | 
 c8     | text                        |           | not null |         |             | extended |              | 
Server: loopback
FDW options: (table_name 'types', engine 'MergeTree')

\d+ clickhouse.arrays;
                                     Foreign table "clickhouse.arrays"
 Column |   Type    | Collation | Nullable | Default | FDW options | Storage  | Stats target | Description 
--------+-----------+-----------+----------+---------+-------------+----------+--------------+-------------
 c1     | integer[] |           | not null |         |             | extended |              | 
 c2     | text[]    |           | not null |         |             | extended |              | 
Server: loopback
FDW options: (table_name 'arrays', engine 'MergeTree')

\d+ clickhouse.tuples;
                                     Foreign table "clickhouse.tuples"
 Column |   Type    | Collation | Nullable | Default | FDW options | Storage  | Stats target | Description 
--------+-----------+-----------+----------+---------+-------------+----------+--------------+-------------
 c1     | smallint  |           | not null |         |             | plain    |              | 
 c2     | text      |           | not null |         |             | extended |              | 
 c3.a   | integer[] |           | not null |         |             | extended |              | 
 c3.b   | integer[] |           | not null |         |             | extended |              | 
 c4     | smallint  |           | not null |         |             | plain    |              | 
Server: loopback
FDW options: (table_name 'tuples', engine 'MergeTree')

\d+ clickhouse.dt64;
                                              Foreign table "clickhouse.dt64"
 Column |            Type             | Collation | Nullable | Default | FDW options | Storage | Stats target | Description 
--------+-----------------------------+-----------+----------+---------+-------------+---------+--------------+-------------
 c1     | integer                     |           | not null |         |             | plain   |              | 
 c2     | timestamp without time zone |           | not null |         |             | plain   |              | 
 c3     | timestamp without time zone |           | not null |         |             | plain   |              | 
 c4     | timestamp without time zone |           | not null |         |             | plain   |              | 
Server: loopback
FDW options: (table_name 'dt64', engine 'MergeTree')

SELECT * FROM clickhouse.ints ORDER BY c1 DESC LIMIT 4;
 c1 | c2 | c3 | c4 | c5 | c6 | c7 | c8 |  c9  | c10  
----+----+----+----+----+----+----+----+------+------
 11 | 12 | 13 | 14 | 15 | 16 | 17 | 18 | 19.1 |     
 10 | 11 | 12 | 13 | 14 | 15 | 16 | 17 | 18.1 |     
  9 | 10 | 11 | 12 | 13 | 14 | 15 | 16 | 17.1 | 18.2
  8 |  9 | 10 | 11 | 12 | 13 | 14 | 15 | 16.1 | 17.2
(4 rows)

SELECT * FROM clickhouse.types ORDER BY c1 LIMIT 2;
     c1     |         c2          |    c3    |  c4   |                  c5                  | c6  |  c7   | c9 |     c8     
------------+---------------------+----------+-------+--------------------------------------+-----+-------+----+------------
 1990-01-01 | 1990-01-01 10:00:00 | number 0 | num 0 | f4bf890f-f9dc-4332-ad5c-0c18e73f28e0 | two | three | 0  | cardinal 0
 1990-01-02 | 1990-01-02 10:01:01 | number 1 | num 1 | f4bf890f-f9dc-4332-ad5c-0c18e73f28e1 | two | three | 1  | cardinal 1
(2 rows)

SELECT * FROM clickhouse.arrays ORDER BY c1 LIMIT 2;
  c1   |       c2        
-------+-----------------
 {0,1} | {'num0','num1'}
 {1,2} | {'num1','num2'}
(2 rows)

SELECT * FROM clickhouse.tuples ORDER BY c1 LIMIT 2;
 c1 |    c2     |  c3.a   |  c3.b   | c4 
----+-----------+---------+---------+----
  0 | (0,'0',1) | {0,1,1} | {0,2,2} |  0
  1 | (1,'1',2) | {1,1,1} | {1,2,2} |  1
(2 rows)

IMPORT FOREIGN SCHEMA "<does not matter>" FROM SERVER loopback_bin INTO clickhouse_bin;
NOTICE:  clickhouse_fdw: ClickHouse <Tuple> type was translated to <TEXT> type, please create composite type and alter the column if needed
\d+ clickhouse_bin.ints;
                                       Foreign table "clickhouse_bin.ints"
 Column |       Type       | Collation | Nullable | Default | FDW options | Storage | Stats target | Description 
--------+------------------+-----------+----------+---------+-------------+---------+--------------+-------------
 c1     | smallint         |           | not null |         |             | plain   |              | 
 c2     | smallint         |           | not null |         |             | plain   |              | 
 c3     | integer          |           | not null |         |             | plain   |              | 
 c4     | bigint           |           | not null |         |             | plain   |              | 
 c5     | smallint         |           | not null |         |             | plain   |              | 
 c6     | integer          |           | not null |         |             | plain   |              | 
 c7     | bigint           |           | not null |         |             | plain   |              | 
 c8     | bigint           |           | not null |         |             | plain   |              | 
 c9     | real             |           | not null |         |             | plain   |              | 
 c10    | double precision |           |          |         |             | plain   |              | 
Server: loopback_bin
FDW options: (table_name 'ints', engine 'MergeTree')

\d+ clickhouse_bin.types;
                                            Foreign table "clickhouse_bin.types"
 Column |            Type             | Collation | Nullable | Default | FDW options | Storage  | Stats target | Description 
--------+-----------------------------+-----------+----------+---------+-------------+----------+--------------+-------------
 c1     | date                        |           | not null |         |             | plain    |              | 
 c2     | timestamp without time zone |           | not null |         |             | plain    |              | 
 c3     | text                        |           | not null |         |             | extended |              | 
 c4     | character varying(5)        |           | not null |         |             | extended |              | 
 c5     | uuid                        |           | not null |         |             | plain    |              | 
 c6     | text                        |           | not null |         |             | extended |              | 
 c7     | text                        |           | not null |         |             | extended |              | 
 c9     | character varying(50)       |           |          |         |             | extended |              | 
 c8     | text                        |           | not null |         |             | extended |              | 
Server: loopback_bin
FDW options: (table_name 'types', engine 'MergeTree')

\d+ clickhouse_bin.arrays;
                                   Foreign table "clickhouse_bin.arrays"
 Column |   Type    | Collation | Nullable | Default | FDW options | Storage  | Stats target | Description 
--------+-----------+-----------+----------+---------+-------------+----------+--------------+-------------
 c1     | integer[] |           | not null |         |             | extended |              | 
 c2     | text[]    |           | not null |         |             | extended |              | 
Server: loopback_bin
FDW options: (table_name 'arrays', engine 'MergeTree')

\d+ clickhouse_bin.tuples;
                                   Foreign table "clickhouse_bin.tuples"
 Column |   Type    | Collation | Nullable | Default | FDW options | Storage  | Stats target | Description 
--------+-----------+-----------+----------+---------+-------------+----------+--------------+-------------
 c1     | smallint  |           | not null |         |             | plain    |              | 
 c2     | text      |           | not null |         |             | extended |              | 
 c3.a   | integer[] |           | not null |         |             | extended |              | 
 c3.b   | integer[] |           | not null |         |             | extended |              | 
 c4     | smallint  |           | not null |         |             | plain    |              | 
Server: loopback_bin
FDW options: (table_name 'tuples', engine 'MergeTree')

\d+ clickhouse_bin.dt64;
                                            Foreign table "clickhouse_bin.dt64"
 Column |            Type             | Collation | Nullable | Default | FDW options | Storage | Stats target | Description 
--------+-----------------------------+-----------+----------+---------+-------------+---------+--------------+-------------
 c1     | integer                     |           | not null |         |             | plain   |              | 
 c2     | timestamp without time zone |           | not null |         |             | plain   |              | 
 c3     | timestamp without time zone |           | not null |         |             | plain   |              | 
 c4     | timestamp without time zone |           | not null |         |             | plain   |              | 
Server: loopback_bin
FDW options: (table_name 'dt64', engine 'MergeTree')

SELECT * FROM clickhouse_bin.ints ORDER BY c1 DESC LIMIT 4;
 c1 | c2 | c3 | c4 | c5 | c6 | c7 | c8 |  c9  | c10  
----+----+----+----+----+----+----+----+------+------
 11 | 12 | 13 | 14 | 15 | 16 | 17 | 18 | 19.1 |     
 10 | 11 | 12 | 13 | 14 | 15 | 16 | 17 | 18.1 |     
  9 | 10 | 11 | 12 | 13 | 14 | 15 | 16 | 17.1 | 18.2
  8 |  9 | 10 | 11 | 12 | 13 | 14 | 15 | 16.1 | 17.2
(4 rows)

SELECT * FROM clickhouse_bin.types ORDER BY c1 LIMIT 2;
     c1     |         c2          |    c3    |  c4   |                  c5                  | c6  |  c7   | c9 |     c8     
------------+---------------------+----------+-------+--------------------------------------+-----+-------+----+------------
 1990-01-01 | 1990-01-01 10:00:00 | number 0 | num 0 | f4bf890f-f9dc-4332-ad5c-0c18e73f28e0 | two | three | 0  | cardinal 0
 1990-01-02 | 1990-01-02 10:01:01 | number 1 | num 1 | f4bf890f-f9dc-4332-ad5c-0c18e73f28e1 | two | three | 1  | cardinal 1
(2 rows)

SELECT * FROM clickhouse_bin.arrays ORDER BY c1 LIMIT 2;
  c1   |     c2      
-------+-------------
 {0,1} | {num0,num1}
 {1,2} | {num1,num2}
(2 rows)

SELECT * FROM clickhouse_bin.tuples ORDER BY c1 LIMIT 2;
 c1 |   c2    |  c3.a   |  c3.b   | c4 
----+---------+---------+---------+----
  0 | (0,0,1) | {0,1,1} | {0,2,2} |  0
  1 | (1,1,2) | {1,1,1} | {1,2,2} |  1
(2 rows)

IMPORT FOREIGN SCHEMA "<does not matter>" LIMIT TO (ints, types) FROM SERVER loopback INTO clickhouse_limit;
\d+ clickhouse_limit.ints;
                                      Foreign table "clickhouse_limit.ints"
 Column |       Type       | Collation | Nullable | Default | FDW options | Storage | Stats target | Description 
--------+------------------+-----------+----------+---------+-------------+---------+--------------+-------------
 c1     | smallint         |           | not null |         |             | plain   |              | 
 c2     | smallint         |           | not null |         |             | plain   |              | 
 c3     | integer          |           | not null |         |             | plain   |              | 
 c4     | bigint           |           | not null |         |             | plain   |              | 
 c5     | smallint         |           | not null |         |             | plain   |              | 
 c6     | integer          |           | not null |         |             | plain   |              | 
 c7     | bigint           |           | not null |         |             | plain   |              | 
 c8     | bigint           |           | not null |         |             | plain   |              | 
 c9     | real             |           | not null |         |             | plain   |              | 
 c10    | double precision |           |          |         |             | plain   |              | 
Server: loopback
FDW options: (table_name 'ints', engine 'MergeTree')

\d+ clickhouse_limit.types;
                                           Foreign table "clickhouse_limit.types"
 Column |            Type             | Collation | Nullable | Default | FDW options | Storage  | Stats target | Description 
--------+-----------------------------+-----------+----------+---------+-------------+----------+--------------+-------------
 c1     | date                        |           | not null |         |             | plain    |              | 
 c2     | timestamp without time zone |           | not null |         |             | plain    |              | 
 c3     | text                        |           | not null |         |             | extended |              | 
 c4     | character varying(5)        |           | not null |         |             | extended |              | 
 c5     | uuid                        |           | not null |         |             | plain    |              | 
 c6     | text                        |           | not null |         |             | extended |              | 
 c7     | text                        |           | not null |         |             | extended |              | 
 c9     | character varying(50)       |           |          |         |             | extended |              | 
 c8     | text                        |           | not null |         |             | extended |              | 
Server: loopback
FDW options: (table_name 'types', engine 'MergeTree')

\d+ clickhouse_limit.arrays;
\d+ clickhouse_limit.tuples;
IMPORT FOREIGN SCHEMA "<does not matter>" EXCEPT (ints, types) FROM SERVER loopback INTO clickhouse_except;
NOTICE:  clickhouse_fdw: ClickHouse <Tuple> type was translated to <TEXT> type, please create composite type and alter the column if needed
\d+ clickhouse_except.ints;
\d+ clickhouse_except.types;
\d+ clickhouse_except.arrays;
                                 Foreign table "clickhouse_except.arrays"
 Column |   Type    | Collation | Nullable | Default | FDW options | Storage  | Stats target | Description 
--------+-----------+-----------+----------+---------+-------------+----------+--------------+-------------
 c1     | integer[] |           | not null |         |             | extended |              | 
 c2     | text[]    |           | not null |         |             | extended |              | 
Server: loopback
FDW options: (table_name 'arrays', engine 'MergeTree')

\d+ clickhouse_except.tuples;
                                 Foreign table "clickhouse_except.tuples"
 Column |   Type    | Collation | Nullable | Default | FDW options | Storage  | Stats target | Description 
--------+-----------+-----------+----------+---------+-------------+----------+--------------+-------------
 c1     | smallint  |           | not null |         |             | plain    |              | 
 c2     | text      |           | not null |         |             | extended |              | 
 c3.a   | integer[] |           | not null |         |             | extended |              | 
 c3.b   | integer[] |           | not null |         |             | extended |              | 
 c4     | smallint  |           | not null |         |             | plain    |              | 
Server: loopback
FDW options: (table_name 'tuples', engine 'MergeTree')

DROP USER MAPPING FOR CURRENT_USER SERVER loopback;
DROP USER MAPPING FOR CURRENT_USER SERVER loopback_bin;
SELECT clickhousedb_raw_query('DROP DATABASE regression');
 clickhousedb_raw_query 
------------------------
 
(1 row)

DROP EXTENSION clickhouse_fdw CASCADE;
NOTICE:  drop cascades to 17 other objects
DETAIL:  drop cascades to server loopback
drop cascades to foreign table clickhouse.arrays
drop cascades to foreign table clickhouse.dt64
drop cascades to foreign table clickhouse.ints
drop cascades to foreign table clickhouse.tuples
drop cascades to foreign table clickhouse.types
drop cascades to foreign table clickhouse_limit.ints
drop cascades to foreign table clickhouse_limit.types
drop cascades to foreign table clickhouse_except.arrays
drop cascades to foreign table clickhouse_except.dt64
drop cascades to foreign table clickhouse_except.tuples
drop cascades to server loopback_bin
drop cascades to foreign table clickhouse_bin.arrays
drop cascades to foreign table clickhouse_bin.dt64
drop cascades to foreign table clickhouse_bin.ints
drop cascades to foreign table clickhouse_bin.tuples
drop cascades to foreign table clickhouse_bin.types
